bigquery/storage/managedwriter/adapt/protoconversion.go

```diff
@@ -0,0 +1,259 @@
+	storagepb.TableFieldSchema_INT64:      descriptorpb.FieldDescriptorProto_TYPE_INT64,
+	storagepb.TableFieldSchema_NUMERIC:    descriptorpb.FieldDescriptorProto_TYPE_BYTES,
+	storagepb.TableFieldSchema_STRING:     descriptorpb.FieldDescriptorProto_TYPE_STRING,
+	storagepb.TableFieldSchema_STRUCT:     descriptorpb.FieldDescriptorProto_TYPE_MESSAGE,
+	storagepb.TableFieldSchema_TIME:       descriptorpb.FieldDescriptorProto_TYPE_INT64,
+	storagepb.TableFieldSchema_TIMESTAMP:  descriptorpb.FieldDescriptorProto_TYPE_INT64,
+}
+
+// For TableFieldSchema OPTIONAL mode, we use the wrapper types to allow for the
+// proper representation of NULL values, as proto3 semantics would just use default value.
+var bqTypeToWrapperMap = map[storagepb.TableFieldSchema_Type]string{
+	storagepb.TableFieldSchema_BIGNUMERIC: ".google.protobuf.BytesValue",
+	storagepb.TableFieldSchema_BOOL:       ".google.protobuf.BoolValue",
+	storagepb.TableFieldSchema_BYTES:      ".google.protobuf.BytesValue",
+	storagepb.TableFieldSchema_DATE:       ".google.protobuf.Int32Value",
+	storagepb.TableFieldSchema_DATETIME:   ".google.protobuf.Int64Value",
+	storagepb.TableFieldSchema_DOUBLE:     ".google.protobuf.DoubleValue",
+	storagepb.TableFieldSchema_GEOGRAPHY:  ".google.protobuf.StringValue",
+	storagepb.TableFieldSchema_INT64:      ".google.protobuf.Int64Value",
+	storagepb.TableFieldSchema_NUMERIC:    ".google.protobuf.BytesValue",
+	storagepb.TableFieldSchema_STRING:     ".google.protobuf.StringValue",
+	storagepb.TableFieldSchema_TIME:       ".google.protobuf.Int64Value",
+	storagepb.TableFieldSchema_TIMESTAMP:  ".google.protobuf.Int64Value",
+}
+
+// filename used by well known types proto
+var wellKnownTypesWrapperName = "google/protobuf/wrappers.proto"
+
+// dependencyCache is used to reduce the number of unique messages we generate by caching based on the tableschema.
+//
+// keys are based on the base64-encoded serialized tableschema value.
+type dependencyCache map[string]protoreflect.Descriptor
+
+func (dm *dependencyCache) get(schema *storagepb.TableSchema) protoreflect.Descriptor {
+	if dm == nil {
+		return nil
+	}
+	b, err := proto.Marshal(schema)
+	if err != nil {
+		return nil
+	}
+	encoded := base64.StdEncoding.EncodeToString(b)
+	if desc, ok := (*dm)[encoded]; ok {
+		return desc
+	}
+	return nil
+}
+
+func (dm *dependencyCache) add(schema *storagepb.TableSchema, descriptor protoreflect.Descriptor) error {
+	if dm == nil {
+		return fmt.Errorf("cache is nil")
+	}
+	b, err := proto.Marshal(schema)
+	if err != nil {
+		return fmt.Errorf("failed to serialize tableschema: %v", err)
+	}
+	encoded := base64.StdEncoding.EncodeToString(b)
+	(*dm)[encoded] = descriptor
+	return nil
+}
+
+// StorageSchemaToDescriptor builds a protoreflect.Descriptor for a given table schema.
+func StorageSchemaToDescriptor(inSchema *storagepb.TableSchema, scope string) (protoreflect.Descriptor, error) {
```

Comment:
If this is all meant to be dynamic do we need to expose this to the user at all? Could we just call this method for them given a table schema? I might just be missing the usecase.

Reply:
Specifically, I'm thinking of cases where the user is reading and writing row data and do want to interact directly with a MessageDescriptor via dynamicpb.  This may warrant more helper functions, but I wanted to get back to that after making progress on the writer itself.